### PR TITLE
Fix Rails/HasAndBelongsToMany Errors

### PR DESCRIPTION
**Pull request by macanudo527**
#### What? Why?

- Helps #6055 

This PR rewrites `has_and_belongs_to_many` relationships to use more maintainable and modifiable `has_many` relationships. The existing join tables have been used for the relationships, so the database does not need to be migrated. This should make it easier in the future for us to modify the relationships or add parameters to them more easily. 

#### What should we test?

No further testing should be needed since these are structural changes.

#### Release notes

Changelog Category: Technical changes

### CODE REVIEW

**Comment by mkllnk** [comment]
Moving to In Dev because it's still a draft. Let us know when it's ready for review.

**Comment by macanudo527** [comment]
@mkllnk This should be all right now.

**Comment by mkllnk** [comment]
Thank you for looking into this. It's an interesting rule. There's a Rails feature and this rule recommending not to use it. I'm not sure if I agree with that rule. Why do we declare models if they are not needed?

I personally would deactivate this rule in our style guide. But let's get the opinions of the other developers as well.

It's great to see how this change would look like in your pull request. Seeing the impact makes it much easier to decide on this, I think.

Please approve this pull request if you are for this pull request:

- [ ] @dacook 
- [ ] @rioug 
- [ ] @jibees 
- [x] @mkllnk 
- [ ] @Matt-Yorkley 
- [ ] @abdellani

File: app/models/distributor_payment_method.rb [hunk 1]
```diff
@@ -5,4 +5,7 @@ class DistributorPaymentMethod < ApplicationRecord
 
   belongs_to :payment_method, class_name: "Spree::PaymentMethod", touch: true
   belongs_to :distributor, class_name: "Enterprise", touch: true
+
+  has_many :order_cycle_distributor_payment_methods, dependent: :destroy
+  has_many :order_cycles, through: :order_cycle_distributor_payment_methods
```
Comment (dacook) [inline app/models/distributor_payment_method.rb:10]: In this case, we're defining relationships that we clearly didn't need before.

File: app/models/enterprise.rb [hunk 2]
```diff
@@ -35,8 +35,9 @@ class Enterprise < ApplicationRecord
   has_many :relationships_as_child, class_name: 'EnterpriseRelationship',
                                     foreign_key: 'child_id',
                                     dependent: :destroy
-  has_and_belongs_to_many :groups, join_table: 'enterprise_groups_enterprises',
-                                   class_name: 'EnterpriseGroup'
+  has_many :enterprise_group_enterprises
+  has_many :groups, through: :enterprise_group_enterprises, source: :enterprise_group,
+                    class_name: 'EnterpriseGroup'
```
Comment (dacook) [inline app/models/enterprise.rb:40]: This results in more code to read and put together in your head.

**Review by dacook — COMMENTED** [review]
Rubocop rules are about standardising the codebase and removing ambiguity. `has_and_belongs_to_many` is a long name that may be hard to conceptualise, so at first it sounds like a good idea to make the code more explicit. 

But this isn't the Rails way, as Maikel points out. The Rails way is to hide complexity with "magic", which has it's own challenges to follow. But whether you agree with it or not, I think we should stick with it.

In addition to this, I've briefly reviewed the changes and added some observations below. I actually find it harder to understand than the `has_and_belongs_to_many` convenience method. And I don't think we should be defining models when we don't need to. 
Less code means less to maintain.

So for multiple reasons I agree that we should disable this style rule.

Thanks Neal for working through this, and being patient while we consider it. It must be frustrating when what appears to be simple code style fixes get stalled.

**Comment by jibees** [comment]
I don't want to add anything to this debate, so I'll let you decide what is best for OFN!

**Comment by macanudo527** [comment]
I can see both sides of the argument. The first reason I made these changes was to learn and familiarize myself with Rails database work. It was good to learn, so no real wasted time.

The second reason is that other programmers were commenting that it saves time later when you might want to add more to the relationship between the two models. I suppose you could want to define more parameters to the relationships already defined. For example, for 'enterprise_group_enterprise', you could name this `group_relationship` and define things like start and end dates as well as a role that enterprise might have in the group. I don't see those being useful now, but I'm just brainstorming.

Having said all that, there really isn't that much stopping us from implementing these changes later when we need them since the database table already exists, and all we are really doing is just defining the relationship less 'magically' in Rails. 

I just thought I would present the opposite argument since everyone seems to be supporting keeping `has_belongs_to_many`.

**Comment by mkllnk** [comment]
Thank you for explaining that side as well. There's a potential time saving later but we may never need those models. And in the meantime, they add maintenance to the code base. So far we got three core developers against this rule, one neutral and two haven't answered. So I guess that's a vote against it. :smile: I'll open another PR for changing the rule.

**Pull request closed without merging**